lib/spack/spack/cmd/uninstall.py

```diff
@@ -147,10 +147,10 @@ def do_uninstall(specs, force):
         try:
             # should work if package is known to spack
             packages.append(item.package)
-        except spack.repository.UnknownPackageError:
+        except spack.repository.UnknownEntityError:
             # The package.py file has gone away -- but still
             # want to uninstall.
-            spack.Package(item).do_uninstall(force=True)
+            spack.Package.uninstall_by_spec(item, force=True)
```

Comment:
Definitely :+1: on this, as it seems far more natural to uninstall by `spec` than by `package`. After all specs are there to represent a concrete installation of something. 

I wonder though if we should consider an issue to be fixed the fact that:
```python
assert spec == spec.package.spec
```
is not always true. I've been bitten by this a couple of times already and usually it's fairly complicated to understand where things go the wrong way. The usual scheme for a bug of this kind is:
```python
spec = ... # The good spec, with the state we want
...
# At some point I need to make a call that takes the package
# as a parameter
foo(spec.package)
...

def foo(pkg):
    ...
    # And within foo I need to make a call
    # to a function that wants a spec
    bar(pkg.spec) # <- Here I may have passed a spec with an inconsistent state
```
What do you think @tgamblin ?